proto/aep-type/aep/type/phone_number.proto

```diff
@@ -0,0 +1,121 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package aep.type;
+
+option cc_enable_arenas = true;
+option go_package = "aep.dev/type";
+option java_multiple_files = true;
+option java_outer_classname = "PhoneNumberProto";
+option java_package = "dev.aep.type";
+option objc_class_prefix = "AEP";
+
+import "buf/validate/validate.proto";
+
+// An object representing a phone number in a way suitable as an API wire
+// format.
+//
+// This representation:
+//
+//  - should not be used for locale-specific formatting of a phone number, such
+//    as "+1 (650) 253-0000 ext. 123"
+//
+//  - is not designed for efficient storage
+//  - may not be suitable for dialing - specialized libraries (see references)
+//    should be used to parse the number for that purpose
+//
+// To do something meaningful with this number, such as format it for various
+// use-cases, you will need a library like
+// https://github.com/google/libphonenumber.
+//
+// For instance, in Java this would be:
+//
+//    ```java
+//    dev.aep.type.PhoneNumber wireProto =
+//        dev.aep.type.PhoneNumber.newBuilder().build();
+//    com.google.i18n.phonenumbers.Phonenumber.PhoneNumber phoneNumber =
+//        PhoneNumberUtil.getInstance().parse(wireProto.getE164Number(), "ZZ");
+//    if (!wireProto.getExtension().isEmpty()) {
+//      phoneNumber.setExtension(wireProto.getExtension());
+//    }
+//    ````
+message PhoneNumber {
+  // An object representing a short code, which is a phone number that is
+  // typically much shorter than regular phone numbers and can be used to
+  // address messages in MMS and SMS systems, as well as for abbreviated dialing
+  // (e.g. "Text 611 to see how many minutes you have remaining on your plan.").
+  //
+  // Short codes are restricted to a region and are not internationally
+  // dialable, which means the same short code can exist in different regions,
+  // with different usage and pricing, even if those regions share the same
+  // country calling code (e.g. US and CA).
+  message ShortCode {
+    // Required. The BCP-47 region code of the location where calls to this
+    // short code can be made, such as "US" and "BB".
+    //
+    // Reference(s):
+    //  - http://www.unicode.org/reports/tr35/#unicode_region_subtag
+    string region_code = 1
+        [(buf.validate.field).string = { pattern: "^([A-Z]{2}|[0-9]{3})$" }];
+
+    // Required. The short code digits, without a leading plus ('+') or country
+    // calling code, e.g. "611".
+    string number = 2 [(buf.validate.field).string = { pattern: "^[0-9]+$" }];
+  }
+
+  // Required.  Either a regular number, or a short code.  New fields may be
+  // added to the oneof below in the future, so clients should ignore phone
+  // numbers for which none of the fields they coded against are set.
+  oneof kind {
```

Comment:
is it worth combining shortcodes and phone numbers into a single enum? it feels like to me that it's valuable to keep those types separate - what if you *only* want to accept standard phone numbers?

Or maybe have both a regular phone number message, and a shortcode message, and a common message that has both in a oneof?

Reply:
So first of all, in case it wasn't clear from context, this is Google's `google.type.PhoneNumber` with no modification to the actual fields.  I assume they went through the effort of modeling it correctly.

What do you mean by combining them into a single enum?  What would that look like?

In general, I think it's fine for an API to document restrictions on its fields (with comments and/or with e.g. protovalidate annotations).  An API might say "only US numbers are accepted" or "short codes are not supported" just like it might say "only fully specified dates with a day, month, and year are supported" or "only ISO-4217 currency codes are supported".

Reply:
I think it's a great point. In those situations, I could imagine that one has two options here:

1. create their own types, with just support for the desired fields, copying the validators.
2. use the existing type, but validate on admission time whether only the subset is populated or not.

I'd personally opt for 1, since 2 comes with a worse UX experience (IMO) where type validation happens farther right from authoring the value type. 

So in this case - I'd imagine most people who want to *just* support a phone number, will use a string, and copy the validator instead. 

unfortunately protobuf doesn't really support type aliases / data types, so having a "phone number" type that is just a string isn't possible. So my point is moot for protobuf at least. So I guess you can just leave this as a comment.